File: discord/theme.js
/* ---------------- discord/theme: visual system ----------------
   Extracted from index.js. Holds the colour palette, glyphs, the embed "brand"
   stamp, and every shared embed builder (success/error/denied/...). The display
   name comes from BOT_NAME (env) so the bot can be branded per server.
   The only outside coupling is Discord's EmbedBuilder plus the bot's avatar/version
   for the brand footer - injected as a lazy client accessor + buildId so this loads
   before the Discord client exists and resolves the avatar at send time.

   Usage:
     const theme = require("./discord/theme")({ getClient: () => client, buildId: BUILD_ID });
     const { brand, successEmbed, NV, GLYPH, ... } = theme;
*/
const { EmbedBuilder } = require("discord.js");

module.exports = function createTheme({ getClient, buildId }) {
  // ---- colour palette (exported as NV for backwards compatibility) ----
  // A bright, candy-vibrant set tuned to pop on Discord's dark theme while keeping
  // each colour's *meaning* (success=green, error/ban=red, warn=amber, info=blue...).
  // Every embed in the bot routes its setColor() through these keys, so this table
  // is the single source of truth for the whole UI.
  const NV = {
    AMBER:       0xFBBF24,   // amber-400  - warnings / neutral highlight
    GOLD:        0xFACC15,   // yellow-400 - economy / caps / winnings
    IRRAD_GREEN: 0x4ADE80,   // green-400  - success / online / cleared
    NCR_TAN:     0x38BDF8,   // sky-400    - neutral accent (bright + friendly)
    LEGION_RED:  0xF43F5E,   // rose-500   - bans / blocks / denied
    RUST_RED:    0xFB7185,   // rose-400   - errors (softer than a ban)
    DEAD_GREY:   0x94A3B8,   // slate-400  - disabled / void
    BLUE_VATS:   0x60A5FA,   // blue-400   - info
    DEEP_BLACK:  0x1E293B,   // slate-800  - deep background accent
    // sweet extra accents - reach for these to add variety (casino, factions, fun)
    PURPLE:      0xA78BFA,   // violet-400
    PINK:        0xF472B6,   // pink-400
    TEAL:        0x2DD4BF,   // teal-400
    CYAN:        0x22D3EE,   // cyan-400
    ORANGE:      0xFB923C,   // orange-400
  };

  /* Ban / IP embed accents. */
  const CLIN = {
    red:   0xF43F5E,   // rose-500   - ban / block / active
    green: 0x4ADE80,   // green-400  - cleared / lifted / no bans
    grey:  0xFBBF24,   // amber-400  - neutral info (lists, checks, connection log)
  };

  // ---- embed builders ----
  const DIVIDER = "────────────────────────────";
  const RULE    = "╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌";
  // Display name stamped on every embed. Set BOT_NAME in .env to skin the bot
  // for your RP ("Mojave Authority", "LSPD Command", ...). Neutral default.
  const BRAND_NAME = process.env.BOT_NAME || "Server Authority";
  // Colourful status glyphs - a splash of colour on every title, status line and
  // list. `dot` stays a plain bullet (used as an inline separator in running text).
  const GLYPH = { ok: "✅", bad: "❌", warn: "⚠️", deny: "🚫", info: "ℹ️", dot: "•", up: "🟢", down: "🔴", caps: "💰", rank: "🏅" };

  // ---- visual system  (clean, minimal embeds - see the reference help-menu look) ----
  function brandIcon() { try { return getClient()?.user?.displayAvatarURL?.({ size: 128 }) ?? null; } catch { return null; } }

  /** Minimal stamp: no author header, no avatar/thumbnail, no default footer, no
      timestamp - just the colour bar, title, and body (the clean help-menu look).
      An explicitly passed footer (a short informative note) is kept as plain text;
      any timestamp a call site set is stripped so every embed stays uniform. */
  function brand(embed, { thumb = false, footer } = {}) {
    if (footer) embed.setFooter(typeof footer === "string" ? { text: footer } : footer);
    try { if (embed.data) delete embed.data.timestamp; } catch {}
    clampEmbed(embed);
    return embed;
  }
  // Keep any embed inside Discord's hard API limits so a long field can never
  // reject the whole message (title 256, desc 4096, field name 256/value 1024).
  function clampEmbed(embed) {
    try {
      const d = embed.data;
      if (d.title) embed.setTitle(String(d.title).slice(0, 256));
      if (d.description) embed.setDescription(String(d.description).slice(0, 4096));
      if (Array.isArray(d.fields)) {
        for (const f of d.fields) {
          if (f.name  && f.name.length  > 256)  f.name  = f.name.slice(0, 253)  + "...";
          if (f.value && f.value.length > 1024) f.value = f.value.slice(0, 1021) + "...";
        }
      }
    } catch {}
    return embed;
  }

  /** Fine-grained progress bar - smooth 1/8-cell fill, e.g. ██████▍░░░░░ */
  const _BAR_FRAC = ["", "▏", "▎", "▍", "▌", "▋", "▊", "▉"];
  function bar(value, max, width = 12) {
    const ratio  = max > 0 ? Math.max(0, Math.min(1, value / max)) : 0;
    // Work in 1/8-cell units and carry a rounded-up fraction into a full block,
    // otherwise a fraction that rounds to 8/8 renders as EMPTY and a higher value
    // can draw a shorter bar than a lower one.
    const eighths = Math.round(ratio * width * 8);
    const full    = Math.floor(eighths / 8);
    const frac    = _BAR_FRAC[eighths % 8] || "";
    const used    = full + (frac ? 1 : 0);
    return "█".repeat(full) + frac + "░".repeat(Math.max(0, width - used));
  }
  /** Labeled meter: `██████▍░░░░░  n/max (p%)` - for dashboards and rosters. */
  function meter(value, max, width = 12) {
    const pct = max > 0 ? Math.round((value / max) * 100) : 0;
    return `\`${bar(value, max, width)}\`  **${value}/${max}** *(${pct}%)*`;
  }
  const pip = (ok) => (ok ? GLYPH.up : GLYPH.down);

  // Fixed-width cell for lining up columns inside a monospace code block.
  const cell = (v, w) => { const s = String(v); return s.length > w ? s.slice(0, w - 1) + "..." : s.padEnd(w); };

  /* Intro line at the top of feature embeds - plain text in the clean style
     (was a blockquote-italic "hero quote" in the old theme). */
  function hero(quoteText) { return String(quoteText); }

  /* Ban / IP embeds: stamp them with the full bot branding (author
     header, avatar, timestamp) + an optional footer - same look as everything else. */
  function clinical(embed, footer) {
    return brand(embed, footer ? { footer } : {});
  }

  // ---- embed builders ----
  function successEmbed(title, description) {
    return brand(new EmbedBuilder().setColor(NV.IRRAD_GREEN)
      .setTitle(`${GLYPH.ok} ${title}`)
      .setDescription(String(description)));
  }
  function errorEmbed(title, description) {
    return brand(new EmbedBuilder().setColor(NV.RUST_RED)
      .setTitle(`${GLYPH.bad} ${title}`)
      .setDescription(String(description)));
  }
  function warningEmbed(title, description) {
    return brand(new EmbedBuilder().setColor(NV.AMBER)
      .setTitle(`${GLYPH.warn} ${title}`)
      .setDescription(String(description)));
  }
  /* Shared "you can't run this" card - one look for every access gate. */
  function deniedEmbed(title, description) {
    return brand(new EmbedBuilder().setColor(NV.LEGION_RED)
      .setTitle(`${GLYPH.deny} ${title}`)
      .setDescription(String(description)));
  }
  function adminOnlyEmbed()  { return deniedEmbed("Admins only", "Only admins can use this command."); }
  function ownerOnlyEmbed()  { return deniedEmbed("Owner only", "Only the bot owner can use this command."); }
  function modOnlyEmbed()    { return deniedEmbed("Mods only", "You need the mod role to use this command."); }
  function factionLeaderOnlyEmbed()   { return deniedEmbed("Faction leaders only", "You need the faction leader role (or mod) to use this."); }
  function factionLeaderStrictEmbed() { return deniedEmbed("Faction leaders only", "Only faction leaders can change ranks."); }
  function policeOnlyEmbed()          { return deniedEmbed("Police only", "You need the Police Officer role to manage warrants."); }
  function blacklistedEmbed(entry) {
    const reason = entry?.reason ? ` Reason: ${entry.reason}` : "";
    return deniedEmbed("Blacklisted", `You are blacklisted from this bot, so no commands will work for you.${reason}`);
  }
  function emptyIdEmbed() {
    return warningEmbed("Need a player name", "Type a player name. The field autocompletes anyone online, and typed names work for offline players too.");
  }
  function rateLimitEmbed() {
    return warningEmbed("Slow down", "Too many commands too fast. Wait a few seconds and try again.");
  }

  return {
    NV, CLIN,
    DIVIDER, RULE, BRAND_NAME, GLYPH,
    brandIcon, brand, clampEmbed,
    bar, meter, pip, cell, hero, clinical,
    successEmbed, errorEmbed, warningEmbed, deniedEmbed,
    adminOnlyEmbed, ownerOnlyEmbed, modOnlyEmbed,
    factionLeaderOnlyEmbed, factionLeaderStrictEmbed, policeOnlyEmbed,
    blacklistedEmbed, emptyIdEmbed, rateLimitEmbed,
  };
};

File: dotnet/src/PavlovBot.Host/Discord/Theme.cs
using Discord;

namespace PavlovBot.Host.Discord;

/// <summary>
/// The visual system: one palette, one set of glyphs, one stamp.
/// </summary>
/// <remarks>
/// Every embed the bot sends routes its colour through here, so this table is the single
/// source of truth for the whole UI. Colours keep their MEANING - green is success, red is
/// a ban, amber is a warning - which is what lets somebody read a feed at a glance without
/// reading the words.
///
/// The clamping in <see cref="Brand"/> is not cosmetic. Discord rejects an embed that
/// exceeds any of its limits, and it rejects the WHOLE MESSAGE - so one long ban reason
/// silently costs you the entire audit entry. Truncating is always better than not sending.
/// </remarks>
public static class Theme
{
    /* ---- palette: a Pip-Boy, not a dashboard ----

       THE NAMES KEEP THEIR MEANINGS AND ONLY THE VALUES MOVED. Every embed in the bot picks
       a colour by what it MEANS - green is cleared, red is a ban, amber is a warning - and a
       reader learns that bar before they learn the words. Rethinking the semantics to suit a
       theme would cost the one thing the palette is for; rethinking the hues costs nothing.

       PIP-BOY GREEN IS THE HOUSE COLOUR because it is the one everybody recognises: a CRT
       phosphor green, pulled up in brightness so it survives Discord's dark background,
       which is dark enough to eat the original.

       The rest are the Fallout set pieces - Vault-Tec's blue and yellow, terminal amber,
       radiation orange, rusted steel - chosen so no two land within a few percent of each
       other at a glance. A palette whose warning and whose error are the same orange is a
       palette that says nothing. */
    public static readonly Color Green = new(0x3C, 0xF2, 0x81);        // Pip-Boy phosphor: success, online, cleared
    public static readonly Color Amber = new(0xFF, 0xB0, 0x00);        // terminal amber: warnings
    public static readonly Color Gold = new(0xF2, 0xC1, 0x4E);         // Vault-Tec yellow: caps, economy
    public static readonly Color Blue = new(0x4F, 0x8F, 0xD6);         // Vault-Tec blue: information
    public static readonly Color Sky = new(0x6F, 0xD4, 0xE0);          // cleanroom cyan: neutral accent
    public static readonly Color BanRed = new(0xD9, 0x33, 0x2B);       // klaxon red: bans, blocks, denied
    public static readonly Color ErrorRed = new(0xE8, 0x6A, 0x4C);     // rust: errors, softer than a ban
    public static readonly Color Grey = new(0x8A, 0x85, 0x77);         // wasteland dust: disabled, void
    public static readonly Color Purple = new(0x9B, 0x7E, 0xC8);
    public static readonly Color Pink = new(0xE0, 0x7A, 0x9B);
    public static readonly Color Teal = new(0x3F, 0xB3, 0xA0);
    public static readonly Color Orange = new(0xE8, 0x8A, 0x2E);       // radiation orange

    // ---- glyphs ----
    public const string Ok = "✅";
    public const string Bad = "❌";
    public const string Warn = "⚠️";
    public const string Deny = "🚫";
    public const string Info = "ℹ️";
    public const string Dot = "•";
    public const string Up = "🟢";
    public const string Down = "🔴";
    /// <summary>Money. Caps, not coins - see <see cref="Lore"/>.</summary>
    public const string Money = Lore.Caps;
    public const string Rank = "🏅";

    /// <summary>A terminal rule, for the places that need a break rather than a heading.</summary>
    public const string Divider = "▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔";

    /// <summary>
    /// The name stamped on every branded embed. <c>BOT_NAME</c> skins the bot per server.
    /// </summary>
    /// <remarks>
    /// IT WAS NEVER READ. <c>.env.example</c> has documented BOT_NAME as "stamped on every
    /// embed" since the port, and nothing set this property or printed it - a skin that
    /// silently did nothing, which is the kind of port gap that only shows up when somebody
    /// sets the variable and waits for a change that never comes. <see cref="Brand"/> now
    /// stamps it, and Program wires the variable in.
    /// </remarks>
    public static string BrandName { get; set; } = "Mojave Authority";

    // Discord's hard limits. Exceeding any one of them rejects the entire message.
    private const int MaxTitle = 256;
    private const int MaxDescription = 4096;
    private const int MaxFieldName = 256;
    private const int MaxFieldValue = 1024;
    private const int MaxFields = 25;

    /// <summary>
    /// Apply the house style and clamp everything to Discord's limits.
    /// </summary>
    /// <remarks>
    /// Deliberately minimal: colour bar, title, body, and the brand name in the footer. No
    /// author header, no thumbnail, no timestamp - the clean look the help menu established.
    /// Any timestamp a call site set is stripped so every embed in the bot looks the same.
    /// </remarks>
    /// <summary>
    /// What a footer says besides the brand: the whole text when it was set without one, the part
    /// after the brand when it was branded already, or null when there is nothing else.
    /// </summary>
    internal static string? FooterAfterBrand(string? text)
    {
        if (string.IsNullOrEmpty(text)) return null;

        var mark = $"{Lore.Rads} {BrandName}";
        if (!text.StartsWith(mark, StringComparison.Ordinal)) return text;

        var rest = text[mark.Length..].TrimStart();
        if (rest.StartsWith(Dot, StringComparison.Ordinal)) rest = rest[Dot.Length..].TrimStart();
        return rest.Length > 0 ? rest : null;
    }

    public static EmbedBuilder Brand(this EmbedBuilder embed, string? footer = null)
    {
        ArgumentNullException.ThrowIfNull(embed);

        /* THE BRAND GOES ON EVERYTHING, which is what makes a wall of embeds read as one
           bot rather than thirty commands. The caller's footer keeps its place after it, so
           "Mojave Authority • Updated 14:02 Eastern" is the shape every board and card
           shares. Composed from the argument, never from the footer already set, so branding
           an embed twice cannot stack the name up. */
        /* THE MARK GOES ON EVERYTHING. One glyph in the footer is what makes a wall of embeds
           read as this server's bot rather than as a generic tool that happens to be in the
           channel - and the footer is the one place it can live without competing with a
           title somebody actually needs to read. */
        /* A FOOTER THE CALLER ALREADY SET IS KEPT, after the brand. Branding used to replace it,
           which silently threw away /health's "build <commit>" line - the one way to tell from
           Discord whether a deploy actually replaced the running process - and /monitor's
           legend. An explicit argument still wins. */
        footer ??= FooterAfterBrand(embed.Footer?.Text);

        var stamp = footer is { Length: > 0 }
            ? $"{Lore.Rads} {BrandName} {Dot} {footer}"
            : $"{Lore.Rads} {BrandName}";

        if (stamp is { Length: > 0 }) embed.WithFooter(Truncate(stamp, 2048));
        embed.Timestamp = null;

        if (embed.Title is { Length: > MaxTitle }) embed.Title = Truncate(embed.Title, MaxTitle);
        if (embed.Description is { Length: > MaxDescription }) embed.Description = Truncate(embed.Description, MaxDescription);

        if (embed.Fields.Count > MaxFields) embed.Fields.RemoveRange(MaxFields, embed.Fields.Count - MaxFields);
        foreach (var field in embed.Fields)
        {
            if (field.Name is { Length: > MaxFieldName }) field.Name = Truncate(field.Name, MaxFieldName);
            if (field.Value?.ToString() is { Length: > MaxFieldValue } value) field.Value = Truncate(value, MaxFieldValue);
        }
        return embed;
    }

    private static string Truncate(string text, int limit) =>
        text.Length <= limit ? text : text[..(limit - 3)] + "...";

    private static EmbedBuilder Base(Color color, string title, string? description) =>
        new EmbedBuilder().WithColor(color).WithTitle(title).WithDescription(description ?? "").Brand();

    public static EmbedBuilder Success(string title, string? description = null) => Base(Green, $"{Ok} {title}", description);
    public static EmbedBuilder Failure(string title, string? description = null) => Base(ErrorRed, $"{Bad} {title}", description);
    public static EmbedBuilder Denied(string title, string? description = null) => Base(BanRed, $"{Deny} {title}", description);
    public static EmbedBuilder Warning(string title, string? description = null) => Base(Amber, $"{Warn} {title}", description);
    public static EmbedBuilder Notice(string title, string? description = null) => Base(Blue, $"{Info} {title}", description);

    /// <summary>A ban or other punishment. Its own colour so it stands out in a busy feed.</summary>
    public static EmbedBuilder Punishment(string title, string? description = null) => Base(BanRed, title, description);

    /// <summary>
    /// Split a long list into embed-sized pages.
    /// </summary>
    /// <remarks>
    /// Joining first and slicing the string would cut a line in half. Accumulating whole
    /// lines means a 400-name roster pages cleanly instead of ending a page mid-name.
    /// </remarks>
    public static IReadOnlyList<string> Paginate(IEnumerable<string> lines, int limit = MaxDescription)
    {
        ArgumentNullException.ThrowIfNull(lines);

        var pages = new List<string>();
        var current = new System.Text.StringBuilder();

        foreach (var line in lines)
        {
            var piece = line.Length > limit ? Truncate(line, limit) : line;
            if (current.Length > 0 && current.Length + piece.Length + 1 > limit)
            {
                pages.Add(current.ToString());
                current.Clear();
            }
            if (current.Length > 0) current.Append('\n');
            current.Append(piece);
        }

        if (current.Length > 0) pages.Add(current.ToString());
        return pages.Count > 0 ? pages : [""];
    }

    /// <summary>
    /// A progress bar for a leaderboard row.
    /// </summary>
    /// <remarks>
    /// BLOCK CHARACTERS RATHER THAN AN EMOJI RUN: they are one cell wide in every client, so
    /// a column of them lines up, and they are what a terminal would have drawn anyway.
    /// </remarks>
    public static string Bar(double value, double max, int width = 12)
    {
        if (max <= 0 || width <= 0) return new string('░', Math.Max(0, width));
        var filled = (int)Math.Round(Math.Clamp(value / max, 0, 1) * width);
        return new string('█', filled) + new string('░', width - filled);
    }

    /// <summary>Discord's relative timestamp, which auto-localises per viewer.</summary>
    /// <remarks>
    /// Preferred over any fixed zone WHEN THE READER IS A PERSON IN DISCORD - each of them
    /// sees it in their own. Fixed Eastern is for logs and files, where there is no viewer
    /// to localise for.
    /// </remarks>
    public static string Relative(DateTimeOffset at) => $"<t:{at.ToUnixTimeSeconds()}:R>";

    public static string Absolute(DateTimeOffset at) => $"<t:{at.ToUnixTimeSeconds()}:f>";
}
